### PR TITLE
perf: Add a query processor to rely on tags_flattened column

### PR DESCRIPTION
**Pull request by fpacifici**
This depends on PR #621 
It adds a Query Processor to manipulate the query and identify simple tags conditions at the top level of the conditions clause.
Those conditions are rewritten in one or more LIKE/NOT LIKE expressions over the tags_flattened column.

Specifically these are the cases conditions are optimized:
- top level conditions. No nested OR        
- the condition has to be either in the form tag[t] = value
- functions referencing the tags as parameters are not take into account except for ifNull.
- Both EQ and NEQ conditions are optimized.

Test plan:
unit test added for the corner cases

### CODE REVIEW

File: snuba/datasets/transactions.py [hunk 1]
```diff
@@ -237,4 +238,5 @@ def get_split_query_spec(self) -> Union[None, ColumnSplitSpec]:
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         return [
             PrewhereProcessor(),
+            NestedFieldConditionOptimizer("tags", "tags_map"),
```
Comment (tkaemming) [inline snuba/datasets/transactions.py]: This will need to be updated to reflect #621.

File: snuba/query/processors/tagsmap.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.datasets.transactions_processor import escape_field
+from snuba.query.types import Condition
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition, is_function
+
+
+TAG_PATTERN = re.compile(r"([a-zA-Z0-9_\.]+)\[([a-zA-Z0-9_\.]+)\]")
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]: Can this be consolidated with `NESTED_COL_EXPR_RE`?

https://github.com/getsentry/snuba/blob/657344be7cb89f95246e95854a765a0e873237ab/snuba/datasets/tags_column_processor.py#L16

File: snuba/query/processors/tagsmap.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    The merged columns looks like:
+    tags_map: "tag1:1|tag2:2"
+
+    This means that a simple EQ/NEQ condition like `tags[tag1] == 1` can be
+    transformed into `tags_map LIKE "%tag1:2%"`
+    """
+
+    def __init__(self, nested_col: str, merged_col: str) -> None:
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]:
```suggestion
    def __init__(self, nested_col: str, flattened_col: str) -> None:
```

File: snuba/query/processors/tagsmap.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        if not is_condition(condition):
+            return None
+        if condition[1] not in ["=", "!="]:
+            return None
+        if not isinstance(condition[2], str):
+            # We can only support literals for now.
+            return None
+        lhr = condition[0]
+
+        # This unpoacks the ifNull function. This is just an optimization to make this class more
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]:
```suggestion
        # This unpacks the ifNull function. This is just an optimization to make this class more
```

File: snuba/query/processors/tagsmap.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        This includes these kind of conditions:
+        - top level conditions. No nested OR
+        - the condition has to be either in the form tag[t] = value
+        - functions referencing the tags as parameters are not taken
+          into account except for ifNull.
+        - Both EQ and NEQ conditions are optimized.
+        """
+        if not is_condition(condition):
+            return None
+        if condition[1] not in ["=", "!="]:
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]:
```suggestion
        if condition[1] not in [Operand.EQ.value, Operand.NEQ.value]:
```

Or is there a reason these operators are not assumed to be reflexive? I noticed this on like 88 below, so seems like it may have been intentional?

File: snuba/query/processors/tagsmap.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            return None
+        if not isinstance(condition[2], str):
+            # We can only support literals for now.
+            return None
+        lhr = condition[0]
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]: What does this stand for?

File: snuba/query/processors/tagsmap.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        if function_expr and function_expr[0] == "ifNull" and len(function_expr[1]) > 0:
+            lhr = function_expr[1][0]
+        if not isinstance(lhr, str):
+            return None
+
+        # Now we have a condition in the form of: ["tags[something]", "=", "a string"]
+        tag = TAG_PATTERN.match(lhr)
+        if tag and tag[1] == self.__nested_col:
+            # tag[0] is the full expression that matches the re.
+            nested_col_index = tag[2]
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]: This isn't really an index, this is more of a key or name.

File: snuba/query/processors/tagsmap.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            return OptimizableCondition(
+                nested_col_index=nested_col_index,
+                operand=Operand.EQ if condition[1] == "=" else Operand.NEQ,
+                value=condition[2],
+            )
+        return None
+
+    def process_query(self, query: Query, request_settings: RequestSettings,) -> None:
```
Comment (tkaemming) [inline snuba/query/processors/tagsmap.py]:
```suggestion
    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
```

I don't understand why `black` doesn't fix these. 😕

**Review by tkaemming — APPROVED** [review]
Some minor things inline, but overall this looks good.

It'd be nice if we could quantify the performance improvement, but I guess that would be challenging right now.  It'd also be nice if in the future, all of this mapping type logic would be consolidated in the column definition, and automatically added to the dataset by virtue of it including one of these columns.

File: snuba/query/processors/tagsmap.py (hunk 2 above)
Comment (fpacifici) [reply snuba/query/processors/tagsmap.py]: Not exactly. This one is supposed to be parametric to the column neame (it should work for any nested column) and just identify the structure, while the NESTED_COL_EXPR_RE is meant to identify tags and contexts so it needs to hardcode `tags` and `contexts`. I can ensure we define the content in the same way though.

File: snuba/query/processors/tagsmap.py (hunk 5 above)
Comment (fpacifici) [reply snuba/query/processors/tagsmap.py]: No reason, I just introduced the enum later and forgot this spot.

File: snuba/query/processors/tagsmap.py (hunk 6 above)
Comment (fpacifici) [reply snuba/query/processors/tagsmap.py]: it is a typo for lhs :D

File: snuba/query/processors/tagsmap.py (hunk 8 above)
Comment (fpacifici) [reply snuba/query/processors/tagsmap.py]: it is worse. If you write (like I did)
```
def process_query(
    self,
    query,
):
```
Which is correct. It will make it wrong by turning it into:
```
def process_query(self, query,)
```

by preserving the comma at the end. I suspect it is not the intended behavior, and that, if I did not add the final comma in the first place it would have been rewritten without.

File: snuba/query/processors/tagsmap.py (hunk 2 above)
Comment (tkaemming) [reply snuba/query/processors/tagsmap.py]: I guess I thought wherever `NESTED_COL_EXPR_RE.match` is used, it could share the same general pattern (like `TAG_PATTERN` defined here) and it could be conditional on what the first match group is.

Comment (fpacifici) [reply snuba/query/processors/tagsmap.py]: ok, done

Comment (tkaemming) [reply snuba/query/processors/tagsmap.py]: 👍 Thanks!
